Mathlib/RingTheory/LocalProperties/Exactness.lean

```diff
@@ -0,0 +1,245 @@
+
+theorem surjective_of_localization
+    (h : ∀ (J : Ideal R) (_ : J.IsMaximal), Function.Surjective (map J.primeCompl f)) :
+    Function.Surjective f :=
+  range_eq_top.mp <| eq_top_of_localization_maximal' <|
+  fun J hJ ↦ (localized'_range_eq_range_localizedMap _ _ _ _ f).trans <| range_eq_top.mpr <| h J hJ
+
+theorem bijective_of_localization
+    (h : ∀ (J : Ideal R) (_ : J.IsMaximal), Function.Bijective (map J.primeCompl f)) :
+    Function.Bijective f :=
+  ⟨injective_of_localization _ fun J hJ => (h J hJ).1,
+  surjective_of_localization _ fun J hJ => (h J hJ).2⟩
+
+theorem exact_of_localization {M₀ M₁ M₂ : Type*} [AddCommGroup M₀] [Module R M₀] [AddCommGroup M₁]
+    [Module R M₁] [AddCommGroup M₂] [Module R M₂] (f : M₀ →ₗ[R] M₁) (g : M₁ →ₗ[R] M₂)
+    (h : ∀ (J : Ideal R) (_ : J.IsMaximal),
+    Function.Exact (map J.primeCompl f) (map J.primeCompl g)) :
+    Function.Exact f g := by
+  simp only [LinearMap.exact_iff] at h ⊢
+  apply eq_of_localization_maximal'
+  intro J hJ
+  unfold localized
+  rw [LinearMap.localized'_range_eq_range_localizedMap _ _ (mkLinearMap J.primeCompl M₀) _,
+    LinearMap.localized'_ker_eq_ker_localizedMap _ _ _ (mkLinearMap J.primeCompl M₂)]
+  exact h J hJ
+
+end localization_maximal
+
+section localization_maximal'
+
+open IsLocalizedModule
+
+variable {R A M N} [CommRing R] [CommRing A] [Algebra R A]
+  [AddCommGroup M] [Module R M] [Module A M] [AddCommGroup N] [Module R N] [Module A N]
+
+variable
+  (Rₚ : ∀ (P : Ideal R) [P.IsMaximal], Type*)
+  [∀ (P : Ideal R) [P.IsMaximal], CommRing (Rₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Algebra R (Rₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], IsLocalization.AtPrime (Rₚ P) P]
+  (Mₚ : ∀ (P : Ideal R) [P.IsMaximal], Type*)
+  [∀ (P : Ideal R) [P.IsMaximal], AddCommGroup (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module R (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module (Rₚ P) (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], IsScalarTower R (Rₚ P) (Mₚ P)]
+  (f : ∀ (P : Ideal R) [P.IsMaximal], M →ₗ[R] Mₚ P)
+  [inst : ∀ (P : Ideal R) [P.IsMaximal], IsLocalizedModule P.primeCompl (f P)]
+  (Nₚ : ∀ (P : Ideal R) [P.IsMaximal], Type*)
+  [∀ (P : Ideal R) [P.IsMaximal], AddCommGroup (Nₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module R (Nₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module (Rₚ P) (Nₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], IsScalarTower R (Rₚ P) (Nₚ P)]
+  (g : ∀ (P : Ideal R) [P.IsMaximal], N →ₗ[R] Nₚ P)
+  [inst : ∀ (P : Ideal R) [P.IsMaximal], IsLocalizedModule P.primeCompl (g P)]
+  (k : M →ₗ[R] N)
+include Rₚ
+
+/-- A variant of `injective_of_localization` that accepts `IsLocalizedModule`.-/
+theorem injective_of_localization'
+    (H : ∀ (P : Ideal R) (_ : P.IsMaximal), Function.Injective (map P.primeCompl (f P) (g P) k)) :
+    Function.Injective k :=
+  ker_eq_bot.mp <| eq_bot_of_localization_maximal Rₚ Mₚ f _ <| fun P hP ↦
+    (localized'_ker_eq_ker_localizedMap (Rₚ P) _ (f P) (g P) k).trans <| ker_eq_bot.mpr <| H P hP
```

Comment:
I think it is easier if you do it the other way round, i.e. first prove the general version, then just plugin in `LocalizedModule` etc. into the general version to obtain the specialized one. Also I think the general version deserves the unprimed name, or even `injective_of_isLocalized` and `injective_of_localized`.

Reply:
(same applies below)